### PR TITLE
feat: SSE Prep/Test Cleanup

### PR DESCRIPTION
**Pull request by JamieSinn**
This enables the test harness to generate SSE specific configs if needed and cleans up the repeated config paths ad infinitum to DRY - and make it much cleaner.

Also sets up a proxy options type to better keep track of all the options that the proxies should/could implement.

### CODE REVIEW

File: harness/helpers/helpers.ts [hunk 1]
```diff
@@ -437,11 +441,21 @@ export class LocalTestClient extends BaseTestClient {
...
         return result
     }
+
+    getValidConfigPath() {
+        return `/client/${this.clientId}/config/v1/server/${this.sdkKey}.json`
+    }
+
+    getValidConfig() {
+        return this.options.enableRealtimeUpdates
+            ? config_sse(getMockServerUrl(), `/client/${this.clientId}/sse`)
```
Comment (ajwootto) [inline harness/helpers/helpers.ts:451]: why not just always send the correct SSE paths in the config?

Comment (JamieSinn) [reply harness/helpers/helpers.ts:451]: Dunno - wanted this to be as minimal breaking as possible and that's what felt like an easy option/safe?

File: harness/features/track.local.test.ts [hunk 2]
```diff
@@ -36,10 +35,8 @@ describe('Track Tests - Local', () => {
             url = getConnectionStringForProxy(sdkName)
 
             scope
-                .get(
-                    `/client/${client.clientId}/config/v1/server/${client.sdkKey}.json`,
-                )
-                .reply(200, config)
+                .get(client.getValidConfigPath())
+                .reply(200, client.getValidConfig)
```
Comment (ajwootto) [inline harness/features/track.local.test.ts:39]: passing the function like this might have issues with binding to the `this` context of the client instance, like if this function is using the `this` keyword in it then it wont work

File: harness/helpers/helpers.ts (hunk 1 above)
Comment (ajwootto) [reply harness/helpers/helpers.ts:451]: the configs every SDK is pulling down are gonna have these SSE properties in them, so might as well make it consistent

File: harness/features/track.local.test.ts (hunk 2 above)
Comment (JamieSinn) [reply harness/features/track.local.test.ts:39]: wait what? can you explain more?

File: harness/helpers/helpers.ts (hunk 1 above)
Comment (JamieSinn) [reply harness/helpers/helpers.ts:451]: kk - will fix

File: harness/features/track.local.test.ts (hunk 2 above)
Comment (ajwootto) [reply harness/features/track.local.test.ts:39]: basically in JS  a function is bound to the context its called from, so if you do
```client.someMethod()```
then within that method `this` refers to the client
but if you do something like 
```
const method = client.someMethod
method()
```
then when invoked `this` will be undefined because its not being called within the same context
to fix it you can either do:
```
const method = client.someMethod.bind(client)
method()
```

which makes it so that the method is always bound to "client"

or you can use an arrow function like

```
const method = () => client.someMethod()
method()
```
to basically "wrap" it while preserving context

you can also write the method as an arrow function on the class itself

```
class Client {
 method = () => {
}
}
```
which then makes it always use `this` of the client

File: harness/features/allFeatures.local.test.ts [hunk 3]
```diff
@@ -34,10 +34,8 @@ describe('allFeatures Tests - Local', () => {
             beforeEach(async () => {
                 testClient = new LocalTestClient(sdkName)
                 scope
-                    .get(
-                        `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                    )
-                    .reply(200, config)
+                    .get(testClient.getValidConfigPath())
+                    .reply(200, testClient.getValidConfig())
```
Comment (ajwootto) [inline harness/features/allFeatures.local.test.ts:38]: you could also just make this whole scope thing a single function on the client like `respondWithConfig(scope)`

File: harness/features/initialize.local.test.ts [hunk 4]
```diff
@@ -310,11 +274,9 @@ describe('Initialize Tests - Local', () => {
         it('uses the new config when etag changes, and flushes existing events', async () => {
             const testClient = new LocalTestClient(sdkName)
             scope
-                .get(
-                    `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                )
+                .get(testClient.getValidConfigPath())
                 .times(1)
-                .reply(200, config, {
+                .reply(200, testClient.getValidConfig, {
```
Comment (ajwootto) [inline harness/features/initialize.local.test.ts:279]: call it here

File: harness/features/multithreading.local.test.ts [hunk 5]
```diff
@@ -38,10 +37,8 @@ describe('Multithreading Tests', () => {
                 testClient = new LocalTestClient(sdkName)
 
                 scope
-                    .get(
-                        `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                    )
-                    .reply(200, config, {
+                    .get(testClient.getValidConfigPath())
+                    .reply(200, testClient.getValidConfig, {
```
Comment (ajwootto) [inline harness/features/multithreading.local.test.ts:41]: call this

File: harness/features/allFeatures.local.test.ts (hunk 3 above)
Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:38]: I was debating that - how would i pass through the scope/whatnot?

File: harness/features/initialize.local.test.ts (hunk 4 above)
Comment (JamieSinn) [reply harness/features/initialize.local.test.ts:279]: missed one

File: harness/features/allFeatures.local.test.ts (hunk 3 above)
Comment (ajwootto) [reply harness/features/allFeatures.local.test.ts:38]: just give it as an argument

Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:38]: isn't js pass by value not by ref?

Comment (ajwootto) [reply harness/features/allFeatures.local.test.ts:38]: or there may be a way to retrieve it from somewhere automatically since the scope is set up automatically by the test environment

Comment (ajwootto) [reply harness/features/allFeatures.local.test.ts:38]: > isn't js pass by value not by ref?
nope

Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:38]: yes - but we also override the flow of responding successfully in a few cases - and wanted to keep that functionality

Comment (ajwootto) [reply harness/features/allFeatures.local.test.ts:38]: yeah  you can still construct the responder manually this would just be a slight convenience for the most common case

File: harness/features/variable.local.test.ts [hunk 6]
```diff
@@ -84,10 +83,8 @@ describe('Variable Tests - Local', () => {
                 // conflicting. This one is used to mock the config that the local client is going to use
                 // locally in all of its methods.
                 scope
-                    .get(
-                        `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                    )
-                    .reply(200, config, {
+                    .get(testClient.getValidConfigPath())
+                    .reply(200, testClient.getValidConfig, {
```
Comment (ajwootto) [inline harness/features/variable.local.test.ts:87]: call this

File: harness/features/allFeatures.local.test.ts (hunk 3 above)
Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:38]: yeah - gonna do a client function to make it cleaner

File: harness/features/allFeatures.local.test.ts [hunk 7]
```diff
@@ -32,34 +32,33 @@ describe('allFeatures Tests - Local', () => {
...
+                .reply(200, testClient.getValidConfig)
             await testClient.createClient(true, {
                 configPollingIntervalMS: 60000,
             })
         })
 
-        it('should return all features for user without custom data', async () => {
-            const featuresResponse = await testClient.callAllFeatures({
-                user_id: 'user3',
+            it('should return all features for user without custom data', async () => {
```
Comment (jonathannorris) [inline harness/features/allFeatures.local.test.ts:51]: why is this tabbed in?

Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:51]: I have no idea, prettier does what prettier wants

File: harness/features/bootstrapping.test.ts [hunk 8]
```diff
@@ -11,12 +10,7 @@ describe('Bootstrapping Tests', () => {
 
         describe('not enabled client', () => {
             beforeEach(async () => {
-                testClient = new LocalTestClient(sdkName)
-                scope
```
Comment (jonathannorris) [inline harness/features/bootstrapping.test.ts:15]: why is this removed?

File: harness/features/bootstrapping.test.ts [hunk 9]
```diff
@@ -43,12 +37,7 @@ describe('Bootstrapping Tests', () => {
 
         describe('initialized client', () => {
             beforeEach(async () => {
-                testClient = new LocalTestClient(sdkName)
-                scope
```
Comment (jonathannorris) [inline harness/features/bootstrapping.test.ts:47]: why is this removed?

File: harness/features/initialize.local.test.ts [hunk 10]
```diff
@@ -243,16 +239,11 @@ describe('Initialize Tests - Local', () => {
 
     it('uses the same config if the refetch fails, after retrying once', async () => {
         const testClient = new LocalTestClient(sdkName)
+        addClientSDKConfigRequestMock(testClient)
 
         scope
```
Comment (jonathannorris) [inline harness/features/initialize.local.test.ts:244]: removing the first 200 response here kinda breaks the point of this test.

File: harness/features/bootstrapping.test.ts (hunk 8 above)
Comment (JamieSinn) [reply harness/features/bootstrapping.test.ts:15]: I'm not sure, but the test didn't break? Might be from my rebase?

File: harness/features/initialize.local.test.ts [hunk 11]
```diff
@@ -356,10 +339,10 @@ describe('Initialize Tests - Local', () => {
         const testClient = new LocalTestClient(sdkName)
         scope
             .get(
-                `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
+                testClient.getValidConfigPath(),
             )
             .times(1)
-            .reply(200, config, {
+            .reply(200, testClient.getValidConfig, {
```
Comment (jonathannorris) [inline harness/features/initialize.local.test.ts:345]:
```suggestion
            .reply(200, testClient.getValidConfig(), {
```

File: harness/features/bootstrapping.test.ts (hunk 8 above)
Comment (JamieSinn) [reply harness/features/bootstrapping.test.ts:15]: Oh, it was from the rebase, it is lower down

File: harness/features/initialize.local.test.ts (hunk 11 above)
Comment (jonathannorris) [reply harness/features/initialize.local.test.ts:345]: would probably just be easier to pull out config at the start of this test into a `config` const rather than keep calling `testClient.getValidConfig()`

File: harness/features/variable.local.test.ts [hunk 12]
```diff
@@ -66,22 +66,20 @@ describe('Variable Tests - Local', () => {
...
-            scope
-                .get(
-                    `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                )
-                .reply(200, config, {
-                    ETag: 'local-var-etag',
-                    'Cf-Ray': 'local-ray-id',
-                    'Last-Modified': lastModifiedDate.toUTCString(),
-                })
+        describe('initialized client', () => {
```
Comment (jonathannorris) [inline harness/features/variable.local.test.ts:69]: why is this tabbed in more?

File: harness/helpers/helpers.ts [hunk 13]
```diff
@@ -273,14 +275,22 @@ class BaseTestClient {
...
     private shouldFailInit = false
+    private options: ProxyClientOptions
 
+    constructor(sdkName: string, scope?: Scope) {
+        super(sdkName)
+        if (scope) {
+            this.setupMockConfig(scope)
+        }
+    }
     async createClient(
```
Comment (jonathannorris) [inline harness/helpers/helpers.ts:286]:
```suggestion
    }
    
    async createClient(
```

File: harness/features/variable.local.test.ts (hunk 12 above)
Comment (JamieSinn) [reply harness/features/variable.local.test.ts:69]: can retry prettier - but I have no idea

File: harness/features/initialize.local.test.ts (hunk 11 above)
Comment (JamieSinn) [reply harness/features/initialize.local.test.ts:345]: I kinda wanted to just make it part of the constructor and have a param to override and use a custom pull

File: harness/features/bootstrapping.test.ts (hunk 8 above)
Comment (JamieSinn) [reply harness/features/bootstrapping.test.ts:15]: The scope param was passed in - so it sets it up in the constructor

File: harness/features/bootstrapping.test.ts (hunk 9 above)
Comment (JamieSinn) [reply harness/features/bootstrapping.test.ts:47]: same as above - it uses the scope param

File: .github/dependabot.yml [hunk 14]
```diff
@@ -1,71 +0,0 @@
-version: 2
-updates:
```
Comment (jonathannorris) [inline .github/dependabot.yml:2]: why did you remove all these?

Comment (JamieSinn) [reply .github/dependabot.yml:2]: Dependabot will still give us security updates- this prevents the nagging PRs

File: harness/features/allFeatures.local.test.ts [hunk 15]
```diff
@@ -31,13 +31,7 @@ describe('allFeatures Tests - Local', () => {
         let testClient: LocalTestClient
 
         beforeEach(async () => {
-            testClient = new LocalTestClient(sdkName)
-            scope
```
Comment (jonathannorris) [inline harness/features/allFeatures.local.test.ts:35]: why is this not setting up the config mock anymore?

Comment (JamieSinn) [reply harness/features/allFeatures.local.test.ts:35]: it's set in the client constructor

File: harness/features/bootstrapping.test.ts (hunk 8 above)
Comment (jonathannorris) [reply harness/features/bootstrapping.test.ts:15]: what constructor? like this this is the code on main right now, is it not setting up/testing for the config mock anymore?

File: harness/helpers/helpers.ts [hunk 16]
```diff
@@ -273,14 +275,22 @@ class BaseTestClient {
 
 export class LocalTestClient extends BaseTestClient {
     private shouldFailInit = false
+    private options: ProxyClientOptions
 
+    constructor(sdkName: string, scope?: Scope) {
+        super(sdkName)
+        if (scope) {
+            this.setupMockConfig(scope)
+        }
```
Comment (JamieSinn) [inline harness/helpers/helpers.ts:284]: @jonathannorris here is the setup for the scope

File: harness/helpers/helpers.ts [hunk 17]
```diff
@@ -415,11 +426,25 @@ export class LocalTestClient extends BaseTestClient {
...
+
+    getValidConfigPath() {
+        return `/client/${this.clientId}/config/v1/server/${this.sdkKey}.json`
+    }
+
+    setupMockConfig(scope: Scope) {
+        scope
+            .get(this.getValidConfigPath())
+            .reply(200, this.getValidConfig())
+    }
```
Comment (JamieSinn) [inline harness/helpers/helpers.ts]: @jonathannorris  and this is where it sets up the mock config

File: harness/features/initialize.local.test.ts (hunk 11 above)
Comment (jonathannorris) [reply harness/features/initialize.local.test.ts:299]: also you are still not calling the function here, `()` are missing.

Comment (JamieSinn) [reply harness/features/initialize.local.test.ts:299]: ah - just didn't push it locally

File: harness/features/variable.local.test.ts [hunk 18]
```diff
@@ -74,10 +73,8 @@ describe('Variable Tests - Local', () => {
             // conflicting. This one is used to mock the config that the local client is going to use
             // locally in all of its methods.
             scope
-                .get(
-                    `/client/${testClient.clientId}/config/v1/server/${testClient.sdkKey}.json`,
-                )
-                .reply(200, config, {
+                .get(testClient.getValidConfigPath())
+                .reply(200, testClient.getValidConfig, {
```
Comment (jonathannorris) [inline harness/features/variable.local.test.ts:77]: Don't you need to call the `getValidConfig` method?

Comment (JamieSinn) [reply harness/features/variable.local.test.ts]: yep - fixed in the most recent push

**Review by jonathannorris — APPROVED** [review]
